# Phase 8: Zerodha Data Fetcher for Bear Market Test (Subset 10)

Downloads historical OHLCV data for F&O eligible stocks (Apr 2020 - Nov 2022).

**Prerequisites:**
- Run `phase8_get_fno_universe.py` first to create the F&O universe list
- Zerodha account with API access

**Output:**
- `data/NIFTY200_Subset10/raw/{SYMBOL}.csv` (one file per stock)
- `data/NIFTY200_Subset10/data_quality_report.txt`

## 1. Setup and Imports

In [1]:
import os
import sys
import time
import logging
import json
from datetime import datetime, timedelta, date
from typing import List, Dict, Optional
import pandas as pd
import numpy as np
from kiteconnect import KiteConnect

# Change to project root directory
os.chdir('/home/ubuntu/rajnish/Multitask-Stockformer')
print(f"Working directory: {os.getcwd()}")

Working directory: /home/ubuntu/rajnish/Multitask-Stockformer


## 2. Configuration

**Update date range and paths as needed:**

In [2]:
# Zerodha API credentials (same as Phase 2)
API_KEY = "a3vlmmcvyt40udoq"
API_SECRET = "xin86nvnojty5996zzexbu7chc040zy0"

# Date range for Subset 10 (Apr 2020 - Nov 2022)
FROM_DATE = "2020-04-01"
TO_DATE = "2022-11-30"

# Paths
OUTPUT_DIR = "./data/NIFTY200_Subset10"
RAW_DATA_DIR = os.path.join(OUTPUT_DIR, "raw")
UNIVERSE_FILE = os.path.join(OUTPUT_DIR, "fno_universe_2020_04.txt")
TOKEN_FILE = os.path.join(OUTPUT_DIR, "zerodha_token.json")
LOG_FILE = os.path.join(OUTPUT_DIR, "zerodha_fetcher.log")

# Data quality thresholds
REQUIRED_COVERAGE = 0.80  # Require 80% coverage of trading days

# Create directories
os.makedirs(OUTPUT_DIR, exist_ok=True)
os.makedirs(RAW_DATA_DIR, exist_ok=True)

print(f"API Key: ✓ Configured")
print(f"Date Range: {FROM_DATE} to {TO_DATE}")
print(f"Output Directory: {OUTPUT_DIR}")

API Key: ✓ Configured
Date Range: 2020-04-01 to 2022-11-30
Output Directory: ./data/NIFTY200_Subset10


## 3. Configure Logging

In [3]:
# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(LOG_FILE),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)
logger.info("Logger configured successfully")

2026-01-11 11:37:56,050 - INFO - Logger configured successfully


## 4. Token Management Functions

In [4]:
def save_token(access_token: str) -> dict:
    """Save access token with timestamp to file"""
    token_data = {
        'access_token': access_token,
        'timestamp': datetime.now().isoformat(),
        'date': datetime.now().date().isoformat()
    }
    
    with open(TOKEN_FILE, 'w') as f:
        json.dump(token_data, f, indent=2)
    
    logger.info(f"Token saved to {TOKEN_FILE}")
    return token_data

def load_token() -> Optional[str]:
    """Load access token if valid (same day)"""
    if not os.path.exists(TOKEN_FILE):
        logger.info("No saved token found")
        return None
    
    try:
        with open(TOKEN_FILE, 'r') as f:
            token_data = json.load(f)
        
        # Check if token is from today
        saved_date = date.fromisoformat(token_data['date'])
        today = datetime.now().date()
        
        if saved_date == today:
            logger.info(f"✓ Found valid token from {token_data['timestamp']}")
            return token_data['access_token']
        else:
            logger.info(f"Token expired (from {saved_date}), need fresh login")
            return None
    
    except Exception as e:
        logger.error(f"Error loading token: {e}")
        return None

print("✓ Token management functions defined")

✓ Token management functions defined


## 5. Initialize Zerodha API Connection

**If token is expired, this cell will show the login URL. Follow the instructions to get a new token.**

In [5]:
# Initialize KiteConnect
kite = KiteConnect(api_key=API_KEY)

# Try to load existing token
access_token = load_token()

if access_token:
    # Use saved token
    kite.set_access_token(access_token)
    print("✅ Using saved session - no login required!")
else:
    # Manual login required
    print("\n" + "="*80)
    print("MANUAL LOGIN REQUIRED (One-time per day)")
    print("="*80)
    print(f"\n1. Open this URL in your browser:\n   {kite.login_url()}")
    print("\n2. After login, you'll be redirected to: http://127.0.0.1/?request_token=...")
    print("\n3. Copy the 'request_token' value and paste it in the next cell")

2026-01-11 11:38:10,817 - INFO - No saved token found



MANUAL LOGIN REQUIRED (One-time per day)

1. Open this URL in your browser:
   https://kite.zerodha.com/connect/login?api_key=a3vlmmcvyt40udoq&v=3

2. After login, you'll be redirected to: http://127.0.0.1/?request_token=...

3. Copy the 'request_token' value and paste it in the next cell


## 6. Complete Login (if needed)

**Only run this cell if the previous cell showed "MANUAL LOGIN REQUIRED"**

Paste your request_token below:

In [6]:
# ONLY RUN THIS IF LOGIN REQUIRED
# Paste your request_token here:
request_token = "erR3eiW4oWxYCJA95QLFMn8KFbKqytLb"  # <-- Paste token here between the quotes

if request_token:
    try:
        data = kite.generate_session(request_token, api_secret=API_SECRET)
        access_token = data["access_token"]
        kite.set_access_token(access_token)
        save_token(access_token)
        print("✅ Login Successful! Token saved for today.")
    except Exception as e:
        print(f"❌ Login Failed: {e}")
else:
    print("ℹ️ No request_token provided. Skip this cell if already logged in.")

2026-01-11 11:39:29,847 - INFO - Token saved to ./data/NIFTY200_Subset10/zerodha_token.json


✅ Login Successful! Token saved for today.


## 7. Test Connection

In [7]:
# Test connection
try:
    profile = kite.profile()
    print(f"\n✅ Connected to Zerodha API")
    print(f"User: {profile['user_name']}")
    print(f"Broker: {profile['broker']}")
    print(f"Email: {profile['email']}")
except Exception as e:
    print(f"❌ Connection failed: {e}")
    print("\nPlease complete the login in the previous cell first.")


✅ Connected to Zerodha API
User: Rajnish Ahuja
Broker: ZERODHA
Email: rajnish.ahuja82@gmail.com


## 8. Load F&O Universe

In [12]:
# Check if universe file exists
if not os.path.exists(UNIVERSE_FILE):
    print(f"❌ Error: Universe file not found: {UNIVERSE_FILE}")
    print("Please run phase8_get_fno_universe.py first!")
else:
    # Load F&O universe
    with open(UNIVERSE_FILE, 'r') as f:
        symbols = [line.strip() for line in f if line.strip()]
    
    print(f"✓ Loaded {len(symbols)} stocks from F&O universe")
    print(f"\nFirst 20: {symbols[:20]}")
    print(f"\nLast 10: {symbols[-10:]}")

✓ Loaded 144 stocks from F&O universe

First 20: ['ACC', 'ADANIENT', 'ADANIPORTS', 'ADANIPOWER', 'AMARAJABAT', 'AMBUJACEM', 'APOLLOHOSP', 'APOLLOTYRE', 'ASHOKLEY', 'ASIANPAINT', 'AUROPHARMA', 'AXISBANK', 'BAJAJ-AUTO', 'BAJAJFINSV', 'BAJFINANCE', 'BALKRISIND', 'BANDHANBNK', 'BANKBARODA', 'BATAINDIA', 'BEL']

Last 10: ['TVSMOTOR', 'UBL', 'UJJIVAN', 'ULTRACEMCO', 'UPL', 'VEDL', 'VOLTAS', 'WIPRO', 'YESBANK', 'ZEEL']


## 9. Fetch Instruments & Apply Symbol Mapping

In [26]:
# Step 1: Fetch NSE instruments list
print("Fetching NSE instruments list...")
instruments = kite.instruments("NSE")
instrument_map = {i['tradingsymbol']: i['instrument_token'] for i in instruments}
print(f"✓ Fetched {len(instruments)} NSE instruments")

# Step 2: Force reload symbol mapping (in case it was updated)
import importlib
import phase8_symbol_mapping
importlib.reload(phase8_symbol_mapping)
from phase8_symbol_mapping import apply_symbol_mapping, SYMBOL_MAPPING, SKIP_SYMBOLS

# Step 3: Apply mapping to get final available_symbols
available_symbols, mapped_symbols, skipped_symbols = apply_symbol_mapping(symbols, instrument_map)

# Summary
print(f"\n{'='*60}")
print(f"SYMBOL MAPPING SUMMARY")
print(f"{'='*60}")
print(f"Original F&O universe (Apr 2020): {len(symbols)} stocks")
print(f"Direct matches on NSE: {len(available_symbols) - len(mapped_symbols)}")
print(f"Mapped (renamed stocks): {len(mapped_symbols)}")
print(f"Skipped (no successor): {len(skipped_symbols)}")
print(f"\n✓ FINAL: {len(available_symbols)} stocks available for download")

if mapped_symbols:
    print(f"\nMapped symbols:")
    for m in mapped_symbols:
        print(f"    {m}")

if skipped_symbols:
    print(f"\n⚠ Skipped: {skipped_symbols}")

Fetching NSE instruments list...


✓ Fetched 9128 NSE instruments

SYMBOL MAPPING SUMMARY
Original F&O universe (Apr 2020): 144 stocks
Direct matches on NSE: 126
Mapped (renamed stocks): 18
Skipped (no successor): 0

✓ FINAL: 144 stocks available for download

Mapped symbols:
    AMARAJABAT -> ARE&M
    CADILAHC -> ZYDUSLIFE
    CENTURYTEX -> ABREL
    EQUITAS -> EQUITASBNK
    GMRINFRA -> GMRAIRPORT
    HDFC -> HDFCBANK
    IBULHSGFIN -> IBULLSLTD
    INFRATEL -> INDUSTOWER
    L&TFH -> LTF
    MCDOWELL-N -> UNITDSPR
    MINDTREE -> LTIM
    MOTHERSUMI -> MOTHERSON
    NIITTECH -> COFORGE
    PEL -> POONAWALLA
    PVR -> PVRINOX
    SRTRANSFIN -> SHRIRAMFIN
    TATAMOTORS -> TMPV
    UJJIVAN -> UJJIVANSFB


## 10. Define Data Fetching Functions

In [27]:
def fetch_historical_data(symbol: str, instrument_token: int,
                         from_date: str, to_date: str, 
                         max_retries: int = 5) -> Optional[pd.DataFrame]:
    """Fetch historical OHLCV data with exponential backoff retry"""
    from_dt = datetime.strptime(from_date, '%Y-%m-%d')
    to_dt = datetime.strptime(to_date, '%Y-%m-%d')
    
    for attempt in range(max_retries):
        try:
            historical_data = kite.historical_data(
                instrument_token=instrument_token,
                from_date=from_dt,
                to_date=to_dt,
                interval='day'
            )
            
            if not historical_data:
                logger.warning(f"No data returned for {symbol}")
                return None
            
            # Convert to DataFrame
            df = pd.DataFrame(historical_data)
            df = df[['date', 'open', 'high', 'low', 'close', 'volume']]
            df.columns = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume']
            df['Date'] = pd.to_datetime(df['Date']).dt.date
            
            logger.info(f"✓ Fetched {len(df)} records for {symbol}")
            return df
            
        except Exception as e:
            wait_time = 1 * (2 ** attempt)
            logger.warning(f"Attempt {attempt + 1}/{max_retries} failed for {symbol}: {e}")
            
            if attempt < max_retries - 1:
                logger.info(f"Retrying in {wait_time} seconds...")
                time.sleep(wait_time)
            else:
                logger.error(f"✗ Failed after {max_retries} attempts: {symbol}")
                return None
    
    return None

def validate_data(df: pd.DataFrame, symbol: str) -> Dict:
    """Validate data completeness and quality"""
    from_dt = datetime.strptime(FROM_DATE, '%Y-%m-%d').date()
    to_dt = datetime.strptime(TO_DATE, '%Y-%m-%d').date()
    
    validation = {
        'symbol': symbol,
        'total_records': len(df),
        'date_range': f"{df['Date'].min()} to {df['Date'].max()}",
        'coverage': 0.0,
        'zero_volume_days': 0,
        'large_gaps': 0,
        'data_quality': 'PASS'
    }
    
    # Calculate expected trading days (rough estimate: 252 days/year)
    total_days = (to_dt - from_dt).days
    expected_trading_days = int(total_days * 252 / 365)
    actual_days = len(df)
    
    validation['coverage'] = actual_days / expected_trading_days if expected_trading_days > 0 else 0
    
    if validation['coverage'] < REQUIRED_COVERAGE:
        validation['data_quality'] = 'FAIL'
        logger.warning(f"{symbol}: Insufficient coverage {validation['coverage']:.1%}")
    
    # Check for zero volume days
    zero_vol = (df['Volume'] == 0).sum()
    validation['zero_volume_days'] = int(zero_vol)
    if zero_vol > len(df) * 0.1:
        validation['data_quality'] = 'WARN'
    
    # Check for large price gaps
    df_sorted = df.sort_values('Date')
    pct_change = df_sorted['Close'].pct_change().abs()
    large_gaps = (pct_change > 0.40).sum()
    validation['large_gaps'] = int(large_gaps)
    if large_gaps > 0:
        validation['data_quality'] = 'WARN'
    
    return validation

print("✓ Data fetching functions defined")

✓ Data fetching functions defined


In [28]:
# TEST: Download ALL stocks but only 1 day to verify
TEST_MODE = True
TEST_DATE = "2022-01-03"

if TEST_MODE:
    print(f"TEST MODE: {len(available_symbols)} stocks for {TEST_DATE}")
    success, failed = 0, []
    
    for idx, symbol in enumerate(available_symbols, 1):
        token = instrument_map.get(symbol)
        if token:
            df = fetch_historical_data(symbol, token, TEST_DATE, TEST_DATE)
            if df is not None and len(df) > 0:
                success += 1
                if idx <= 3 or idx % 50 == 0:
                    print(f"[{idx}] ✓ {symbol}")
            else:
                failed.append(symbol)
        else:
            failed.append(symbol)
        time.sleep(0.3)
    
    print(f"\n✓ {success}/{len(available_symbols)} successful")
    if failed: print(f"Failed: {failed}")

TEST MODE: 144 stocks for 2022-01-03


2026-01-11 12:23:27,698 - INFO - ✓ Fetched 1 records for ACC


[1] ✓ ACC


2026-01-11 12:23:28,481 - INFO - ✓ Fetched 1 records for ADANIENT


[2] ✓ ADANIENT


2026-01-11 12:23:29,050 - INFO - ✓ Fetched 1 records for ADANIPORTS


[3] ✓ ADANIPORTS


2026-01-11 12:23:29,617 - INFO - ✓ Fetched 1 records for ADANIPOWER
2026-01-11 12:23:30,182 - INFO - ✓ Fetched 1 records for ARE&M
2026-01-11 12:23:30,750 - INFO - ✓ Fetched 1 records for AMBUJACEM
2026-01-11 12:23:32,212 - INFO - ✓ Fetched 1 records for APOLLOHOSP
2026-01-11 12:23:32,774 - INFO - ✓ Fetched 1 records for APOLLOTYRE
2026-01-11 12:23:33,382 - INFO - ✓ Fetched 1 records for ASHOKLEY
2026-01-11 12:23:33,951 - INFO - ✓ Fetched 1 records for ASIANPAINT
2026-01-11 12:23:34,519 - INFO - ✓ Fetched 1 records for AUROPHARMA
2026-01-11 12:23:35,090 - INFO - ✓ Fetched 1 records for AXISBANK
2026-01-11 12:23:35,658 - INFO - ✓ Fetched 1 records for BAJAJ-AUTO
2026-01-11 12:23:36,230 - INFO - ✓ Fetched 1 records for BAJAJFINSV
2026-01-11 12:23:36,799 - INFO - ✓ Fetched 1 records for BAJFINANCE
2026-01-11 12:23:37,365 - INFO - ✓ Fetched 1 records for BALKRISIND
2026-01-11 12:23:37,934 - INFO - ✓ Fetched 1 records for BANDHANBNK
2026-01-11 12:23:38,502 - INFO - ✓ Fetched 1 records for B

[50] ✓ GMRAIRPORT


2026-01-11 12:23:57,420 - INFO - ✓ Fetched 1 records for GODREJCP
2026-01-11 12:23:57,989 - INFO - ✓ Fetched 1 records for GODREJPROP
2026-01-11 12:23:58,559 - INFO - ✓ Fetched 1 records for GRASIM
2026-01-11 12:23:59,122 - INFO - ✓ Fetched 1 records for HAVELLS
2026-01-11 12:23:59,688 - INFO - ✓ Fetched 1 records for HCLTECH
2026-01-11 12:24:00,256 - INFO - ✓ Fetched 1 records for HDFCBANK
2026-01-11 12:24:00,822 - INFO - ✓ Fetched 1 records for HDFCBANK
2026-01-11 12:24:01,389 - INFO - ✓ Fetched 1 records for HDFCLIFE
2026-01-11 12:24:01,954 - INFO - ✓ Fetched 1 records for HEROMOTOCO
2026-01-11 12:24:02,521 - INFO - ✓ Fetched 1 records for HINDALCO
2026-01-11 12:24:03,128 - INFO - ✓ Fetched 1 records for HINDPETRO
2026-01-11 12:24:03,699 - INFO - ✓ Fetched 1 records for HINDUNILVR
2026-01-11 12:24:04,303 - INFO - ✓ Fetched 1 records for IBULLSLTD
2026-01-11 12:24:04,869 - INFO - ✓ Fetched 1 records for ICICIBANK
2026-01-11 12:24:05,435 - INFO - ✓ Fetched 1 records for ICICIPRULI
202

[100] ✓ COFORGE


2026-01-11 12:24:25,971 - INFO - ✓ Fetched 1 records for NMDC
2026-01-11 12:24:26,536 - INFO - ✓ Fetched 1 records for NTPC
2026-01-11 12:24:27,112 - INFO - ✓ Fetched 1 records for OIL
2026-01-11 12:24:27,717 - INFO - ✓ Fetched 1 records for ONGC
2026-01-11 12:24:28,286 - INFO - ✓ Fetched 1 records for PAGEIND
2026-01-11 12:24:29,070 - INFO - ✓ Fetched 1 records for POONAWALLA
2026-01-11 12:24:29,639 - INFO - ✓ Fetched 1 records for PETRONET
2026-01-11 12:24:30,209 - INFO - ✓ Fetched 1 records for PFC
2026-01-11 12:24:30,776 - INFO - ✓ Fetched 1 records for PIDILITIND
2026-01-11 12:24:31,380 - INFO - ✓ Fetched 1 records for PNB
2026-01-11 12:24:31,986 - INFO - ✓ Fetched 1 records for POWERGRID
2026-01-11 12:24:32,552 - INFO - ✓ Fetched 1 records for PVRINOX
2026-01-11 12:24:33,158 - INFO - ✓ Fetched 1 records for RAMCOCEM
2026-01-11 12:24:33,764 - INFO - ✓ Fetched 1 records for RBLBANK
2026-01-11 12:24:34,327 - INFO - ✓ Fetched 1 records for RECLTD
2026-01-11 12:24:34,896 - INFO - ✓ Fe


✓ 144/144 successful


## 11. Download All Data

**This will take ~1-2 hours for 144 stocks due to API rate limits.**

In [29]:
# Initialize tracking
validation_results = []
successful = 0
failed = []

print("="*80)
print(f"Starting data download for {len(available_symbols)} stocks")
print(f"Date range: {FROM_DATE} to {TO_DATE}")
print("="*80 + "\n")

# Download data for each symbol
for idx, symbol in enumerate(available_symbols, 1):
    print(f"\n[{idx}/{len(available_symbols)}] Processing {symbol}...")
    
    instrument_token = instrument_map[symbol]
    
    # Fetch data
    df = fetch_historical_data(symbol, instrument_token, FROM_DATE, TO_DATE)
    
    if df is not None and len(df) > 0:
        # Validate data
        validation = validate_data(df, symbol)
        validation_results.append(validation)
        
        # Save to CSV only if passes quality check
        if validation['data_quality'] != 'FAIL':
            output_file = os.path.join(RAW_DATA_DIR, f"{symbol}.csv")
            df.to_csv(output_file, index=False)
            print(f"✓ Saved {len(df)} records to {output_file}")
            successful += 1
        else:
            print(f"⚠ Skipped {symbol} due to insufficient data coverage")
            failed.append(symbol)
        
        # Rate limiting (Zerodha API limit)
        time.sleep(0.5)
    else:
        failed.append(symbol)
        print(f"✗ Failed to download {symbol}")

print("\n" + "="*80)
print("DOWNLOAD COMPLETE")
print("="*80)
print(f"Successful: {successful}/{len(available_symbols)}")
print(f"Failed: {len(failed)}")
if failed:
    print(f"\nFailed symbols: {', '.join(failed[:20])}{'...' if len(failed) > 20 else ''}")

Starting data download for 144 stocks
Date range: 2020-04-01 to 2022-11-30


[1/144] Processing ACC...


2026-01-11 12:45:11,958 - INFO - ✓ Fetched 662 records for ACC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ACC.csv

[2/144] Processing ADANIENT...


2026-01-11 12:45:12,844 - INFO - ✓ Fetched 662 records for ADANIENT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ADANIENT.csv

[3/144] Processing ADANIPORTS...


2026-01-11 12:45:13,702 - INFO - ✓ Fetched 662 records for ADANIPORTS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ADANIPORTS.csv

[4/144] Processing ADANIPOWER...


2026-01-11 12:45:14,562 - INFO - ✓ Fetched 662 records for ADANIPOWER


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ADANIPOWER.csv

[5/144] Processing ARE&M...


2026-01-11 12:45:15,416 - INFO - ✓ Fetched 662 records for ARE&M


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ARE&M.csv

[6/144] Processing AMBUJACEM...


2026-01-11 12:45:16,266 - INFO - ✓ Fetched 662 records for AMBUJACEM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/AMBUJACEM.csv

[7/144] Processing APOLLOHOSP...


2026-01-11 12:45:17,127 - INFO - ✓ Fetched 662 records for APOLLOHOSP


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/APOLLOHOSP.csv

[8/144] Processing APOLLOTYRE...


2026-01-11 12:45:17,984 - INFO - ✓ Fetched 662 records for APOLLOTYRE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/APOLLOTYRE.csv

[9/144] Processing ASHOKLEY...


2026-01-11 12:45:18,835 - INFO - ✓ Fetched 662 records for ASHOKLEY


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ASHOKLEY.csv

[10/144] Processing ASIANPAINT...


2026-01-11 12:45:19,677 - INFO - ✓ Fetched 662 records for ASIANPAINT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ASIANPAINT.csv

[11/144] Processing AUROPHARMA...


2026-01-11 12:45:20,537 - INFO - ✓ Fetched 662 records for AUROPHARMA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/AUROPHARMA.csv

[12/144] Processing AXISBANK...


2026-01-11 12:45:21,398 - INFO - ✓ Fetched 662 records for AXISBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/AXISBANK.csv

[13/144] Processing BAJAJ-AUTO...


2026-01-11 12:45:22,257 - INFO - ✓ Fetched 662 records for BAJAJ-AUTO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BAJAJ-AUTO.csv

[14/144] Processing BAJAJFINSV...


2026-01-11 12:45:23,111 - INFO - ✓ Fetched 662 records for BAJAJFINSV


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BAJAJFINSV.csv

[15/144] Processing BAJFINANCE...


2026-01-11 12:45:23,964 - INFO - ✓ Fetched 662 records for BAJFINANCE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BAJFINANCE.csv

[16/144] Processing BALKRISIND...


2026-01-11 12:45:24,823 - INFO - ✓ Fetched 662 records for BALKRISIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BALKRISIND.csv

[17/144] Processing BANDHANBNK...


2026-01-11 12:45:25,673 - INFO - ✓ Fetched 662 records for BANDHANBNK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BANDHANBNK.csv

[18/144] Processing BANKBARODA...


2026-01-11 12:45:26,542 - INFO - ✓ Fetched 662 records for BANKBARODA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BANKBARODA.csv

[19/144] Processing BATAINDIA...


2026-01-11 12:45:27,394 - INFO - ✓ Fetched 662 records for BATAINDIA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BATAINDIA.csv

[20/144] Processing BEL...


2026-01-11 12:45:28,256 - INFO - ✓ Fetched 662 records for BEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BEL.csv

[21/144] Processing BERGEPAINT...


2026-01-11 12:45:29,119 - INFO - ✓ Fetched 662 records for BERGEPAINT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BERGEPAINT.csv

[22/144] Processing BHARATFORG...


2026-01-11 12:45:30,179 - INFO - ✓ Fetched 662 records for BHARATFORG


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BHARATFORG.csv

[23/144] Processing BHARTIARTL...


2026-01-11 12:45:31,030 - INFO - ✓ Fetched 662 records for BHARTIARTL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BHARTIARTL.csv

[24/144] Processing BHEL...


2026-01-11 12:45:31,886 - INFO - ✓ Fetched 662 records for BHEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BHEL.csv

[25/144] Processing BIOCON...


2026-01-11 12:45:32,735 - INFO - ✓ Fetched 662 records for BIOCON


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BIOCON.csv

[26/144] Processing BOSCHLTD...


2026-01-11 12:45:33,592 - INFO - ✓ Fetched 662 records for BOSCHLTD


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BOSCHLTD.csv

[27/144] Processing BPCL...


2026-01-11 12:45:34,468 - INFO - ✓ Fetched 662 records for BPCL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BPCL.csv

[28/144] Processing BRITANNIA...


2026-01-11 12:45:35,312 - INFO - ✓ Fetched 662 records for BRITANNIA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/BRITANNIA.csv

[29/144] Processing ZYDUSLIFE...


2026-01-11 12:45:36,182 - INFO - ✓ Fetched 662 records for ZYDUSLIFE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ZYDUSLIFE.csv

[30/144] Processing CANBK...


2026-01-11 12:45:37,034 - INFO - ✓ Fetched 662 records for CANBK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CANBK.csv

[31/144] Processing ABREL...


2026-01-11 12:45:37,888 - INFO - ✓ Fetched 662 records for ABREL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ABREL.csv

[32/144] Processing CESC...


2026-01-11 12:45:38,752 - INFO - ✓ Fetched 662 records for CESC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CESC.csv

[33/144] Processing CHOLAFIN...


2026-01-11 12:45:39,606 - INFO - ✓ Fetched 662 records for CHOLAFIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CHOLAFIN.csv

[34/144] Processing CIPLA...


2026-01-11 12:45:40,463 - INFO - ✓ Fetched 662 records for CIPLA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CIPLA.csv

[35/144] Processing COALINDIA...


2026-01-11 12:45:41,324 - INFO - ✓ Fetched 662 records for COALINDIA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/COALINDIA.csv

[36/144] Processing COLPAL...


2026-01-11 12:45:42,333 - INFO - ✓ Fetched 662 records for COLPAL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/COLPAL.csv

[37/144] Processing CONCOR...


2026-01-11 12:45:43,190 - INFO - ✓ Fetched 662 records for CONCOR


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CONCOR.csv

[38/144] Processing CUMMINSIND...


2026-01-11 12:45:44,046 - INFO - ✓ Fetched 662 records for CUMMINSIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/CUMMINSIND.csv

[39/144] Processing DABUR...


2026-01-11 12:45:44,895 - INFO - ✓ Fetched 662 records for DABUR


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/DABUR.csv

[40/144] Processing DIVISLAB...


2026-01-11 12:45:45,741 - INFO - ✓ Fetched 662 records for DIVISLAB


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/DIVISLAB.csv

[41/144] Processing DLF...


2026-01-11 12:45:46,600 - INFO - ✓ Fetched 662 records for DLF


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/DLF.csv

[42/144] Processing DRREDDY...


2026-01-11 12:45:47,456 - INFO - ✓ Fetched 662 records for DRREDDY


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/DRREDDY.csv

[43/144] Processing EICHERMOT...


2026-01-11 12:45:48,318 - INFO - ✓ Fetched 662 records for EICHERMOT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/EICHERMOT.csv

[44/144] Processing EQUITASBNK...


2026-01-11 12:45:49,150 - INFO - ✓ Fetched 516 records for EQUITASBNK
2026-01-11 12:45:49,152 - WARNING - EQUITASBNK: Insufficient coverage 76.9%


⚠ Skipped EQUITASBNK due to insufficient data coverage

[45/144] Processing ESCORTS...


2026-01-11 12:45:50,015 - INFO - ✓ Fetched 662 records for ESCORTS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ESCORTS.csv

[46/144] Processing EXIDEIND...


2026-01-11 12:45:50,878 - INFO - ✓ Fetched 662 records for EXIDEIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/EXIDEIND.csv

[47/144] Processing FEDERALBNK...


2026-01-11 12:45:51,735 - INFO - ✓ Fetched 662 records for FEDERALBNK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/FEDERALBNK.csv

[48/144] Processing GAIL...


2026-01-11 12:45:52,588 - INFO - ✓ Fetched 662 records for GAIL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GAIL.csv

[49/144] Processing GLENMARK...


2026-01-11 12:45:53,436 - INFO - ✓ Fetched 662 records for GLENMARK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GLENMARK.csv

[50/144] Processing GMRAIRPORT...


2026-01-11 12:45:54,307 - INFO - ✓ Fetched 662 records for GMRAIRPORT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GMRAIRPORT.csv

[51/144] Processing GODREJCP...


2026-01-11 12:45:55,152 - INFO - ✓ Fetched 662 records for GODREJCP


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GODREJCP.csv

[52/144] Processing GODREJPROP...


2026-01-11 12:45:56,002 - INFO - ✓ Fetched 662 records for GODREJPROP


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GODREJPROP.csv

[53/144] Processing GRASIM...


2026-01-11 12:45:56,851 - INFO - ✓ Fetched 662 records for GRASIM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/GRASIM.csv

[54/144] Processing HAVELLS...


2026-01-11 12:45:57,714 - INFO - ✓ Fetched 662 records for HAVELLS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HAVELLS.csv

[55/144] Processing HCLTECH...


2026-01-11 12:45:58,570 - INFO - ✓ Fetched 662 records for HCLTECH


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HCLTECH.csv

[56/144] Processing HDFCBANK...


2026-01-11 12:45:59,426 - INFO - ✓ Fetched 662 records for HDFCBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HDFCBANK.csv

[57/144] Processing HDFCBANK...


2026-01-11 12:46:00,274 - INFO - ✓ Fetched 662 records for HDFCBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HDFCBANK.csv

[58/144] Processing HDFCLIFE...


2026-01-11 12:46:01,128 - INFO - ✓ Fetched 662 records for HDFCLIFE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HDFCLIFE.csv

[59/144] Processing HEROMOTOCO...


2026-01-11 12:46:01,985 - INFO - ✓ Fetched 662 records for HEROMOTOCO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HEROMOTOCO.csv

[60/144] Processing HINDALCO...


2026-01-11 12:46:02,834 - INFO - ✓ Fetched 662 records for HINDALCO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HINDALCO.csv

[61/144] Processing HINDPETRO...


2026-01-11 12:46:03,690 - INFO - ✓ Fetched 662 records for HINDPETRO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HINDPETRO.csv

[62/144] Processing HINDUNILVR...


2026-01-11 12:46:04,550 - INFO - ✓ Fetched 662 records for HINDUNILVR


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/HINDUNILVR.csv

[63/144] Processing IBULLSLTD...


2026-01-11 12:46:05,397 - INFO - ✓ Fetched 662 records for IBULLSLTD


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/IBULLSLTD.csv

[64/144] Processing ICICIBANK...


2026-01-11 12:46:06,461 - INFO - ✓ Fetched 662 records for ICICIBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ICICIBANK.csv

[65/144] Processing ICICIPRULI...


2026-01-11 12:46:07,310 - INFO - ✓ Fetched 662 records for ICICIPRULI


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ICICIPRULI.csv

[66/144] Processing IDEA...


2026-01-11 12:46:08,162 - INFO - ✓ Fetched 662 records for IDEA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/IDEA.csv

[67/144] Processing IDFCFIRSTB...


2026-01-11 12:46:09,035 - INFO - ✓ Fetched 662 records for IDFCFIRSTB


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/IDFCFIRSTB.csv

[68/144] Processing IGL...


2026-01-11 12:46:09,896 - INFO - ✓ Fetched 662 records for IGL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/IGL.csv

[69/144] Processing INDIGO...


2026-01-11 12:46:10,774 - INFO - ✓ Fetched 662 records for INDIGO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/INDIGO.csv

[70/144] Processing INDUSINDBK...


2026-01-11 12:46:11,698 - INFO - ✓ Fetched 662 records for INDUSINDBK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/INDUSINDBK.csv

[71/144] Processing INDUSTOWER...


2026-01-11 12:46:12,553 - INFO - ✓ Fetched 662 records for INDUSTOWER


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/INDUSTOWER.csv

[72/144] Processing INFY...


2026-01-11 12:46:13,413 - INFO - ✓ Fetched 662 records for INFY


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/INFY.csv

[73/144] Processing IOC...


2026-01-11 12:46:14,284 - INFO - ✓ Fetched 662 records for IOC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/IOC.csv

[74/144] Processing ITC...


2026-01-11 12:46:15,134 - INFO - ✓ Fetched 662 records for ITC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ITC.csv

[75/144] Processing JINDALSTEL...


2026-01-11 12:46:15,993 - INFO - ✓ Fetched 662 records for JINDALSTEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/JINDALSTEL.csv

[76/144] Processing JSWSTEEL...


2026-01-11 12:46:16,847 - INFO - ✓ Fetched 662 records for JSWSTEEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/JSWSTEEL.csv

[77/144] Processing JUBLFOOD...


2026-01-11 12:46:17,699 - INFO - ✓ Fetched 662 records for JUBLFOOD


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/JUBLFOOD.csv

[78/144] Processing JUSTDIAL...


2026-01-11 12:46:18,543 - INFO - ✓ Fetched 662 records for JUSTDIAL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/JUSTDIAL.csv

[79/144] Processing KOTAKBANK...


2026-01-11 12:46:19,386 - INFO - ✓ Fetched 662 records for KOTAKBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/KOTAKBANK.csv

[80/144] Processing LTF...


2026-01-11 12:46:20,247 - INFO - ✓ Fetched 662 records for LTF


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/LTF.csv

[81/144] Processing LICHSGFIN...


2026-01-11 12:46:21,117 - INFO - ✓ Fetched 662 records for LICHSGFIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/LICHSGFIN.csv

[82/144] Processing LT...


2026-01-11 12:46:21,971 - INFO - ✓ Fetched 662 records for LT


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/LT.csv

[83/144] Processing LUPIN...


2026-01-11 12:46:22,836 - INFO - ✓ Fetched 662 records for LUPIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/LUPIN.csv

[84/144] Processing M&M...


2026-01-11 12:46:23,680 - INFO - ✓ Fetched 662 records for M&M


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/M&M.csv

[85/144] Processing M&MFIN...


2026-01-11 12:46:24,539 - INFO - ✓ Fetched 662 records for M&MFIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/M&MFIN.csv

[86/144] Processing MANAPPURAM...


2026-01-11 12:46:25,394 - INFO - ✓ Fetched 662 records for MANAPPURAM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MANAPPURAM.csv

[87/144] Processing MARICO...


2026-01-11 12:46:26,265 - INFO - ✓ Fetched 662 records for MARICO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MARICO.csv

[88/144] Processing MARUTI...


2026-01-11 12:46:27,122 - INFO - ✓ Fetched 662 records for MARUTI


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MARUTI.csv

[89/144] Processing UNITDSPR...


2026-01-11 12:46:27,967 - INFO - ✓ Fetched 662 records for UNITDSPR


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/UNITDSPR.csv

[90/144] Processing MFSL...


2026-01-11 12:46:28,816 - INFO - ✓ Fetched 662 records for MFSL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MFSL.csv

[91/144] Processing MGL...


2026-01-11 12:46:29,671 - INFO - ✓ Fetched 662 records for MGL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MGL.csv

[92/144] Processing LTIM...


2026-01-11 12:46:30,545 - INFO - ✓ Fetched 662 records for LTIM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/LTIM.csv

[93/144] Processing MOTHERSON...


2026-01-11 12:46:31,441 - INFO - ✓ Fetched 662 records for MOTHERSON


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MOTHERSON.csv

[94/144] Processing MRF...


2026-01-11 12:46:32,297 - INFO - ✓ Fetched 662 records for MRF


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MRF.csv

[95/144] Processing MUTHOOTFIN...


2026-01-11 12:46:33,149 - INFO - ✓ Fetched 662 records for MUTHOOTFIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/MUTHOOTFIN.csv

[96/144] Processing NATIONALUM...


2026-01-11 12:46:33,992 - INFO - ✓ Fetched 662 records for NATIONALUM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NATIONALUM.csv

[97/144] Processing NAUKRI...


2026-01-11 12:46:34,851 - INFO - ✓ Fetched 662 records for NAUKRI


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NAUKRI.csv

[98/144] Processing NCC...


2026-01-11 12:46:35,704 - INFO - ✓ Fetched 662 records for NCC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NCC.csv

[99/144] Processing NESTLEIND...


2026-01-11 12:46:36,566 - INFO - ✓ Fetched 662 records for NESTLEIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NESTLEIND.csv

[100/144] Processing COFORGE...


2026-01-11 12:46:37,413 - INFO - ✓ Fetched 662 records for COFORGE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/COFORGE.csv

[101/144] Processing NMDC...


2026-01-11 12:46:38,258 - INFO - ✓ Fetched 662 records for NMDC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NMDC.csv

[102/144] Processing NTPC...


2026-01-11 12:46:39,105 - INFO - ✓ Fetched 662 records for NTPC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/NTPC.csv

[103/144] Processing OIL...


2026-01-11 12:46:39,953 - INFO - ✓ Fetched 662 records for OIL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/OIL.csv

[104/144] Processing ONGC...


2026-01-11 12:46:40,806 - INFO - ✓ Fetched 662 records for ONGC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ONGC.csv

[105/144] Processing PAGEIND...


2026-01-11 12:46:41,687 - INFO - ✓ Fetched 662 records for PAGEIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PAGEIND.csv

[106/144] Processing POONAWALLA...


2026-01-11 12:46:42,540 - INFO - ✓ Fetched 662 records for POONAWALLA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/POONAWALLA.csv

[107/144] Processing PETRONET...


2026-01-11 12:46:43,391 - INFO - ✓ Fetched 662 records for PETRONET


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PETRONET.csv

[108/144] Processing PFC...


2026-01-11 12:46:44,249 - INFO - ✓ Fetched 662 records for PFC


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PFC.csv

[109/144] Processing PIDILITIND...


2026-01-11 12:46:45,118 - INFO - ✓ Fetched 662 records for PIDILITIND


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PIDILITIND.csv

[110/144] Processing PNB...


2026-01-11 12:46:45,972 - INFO - ✓ Fetched 662 records for PNB


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PNB.csv

[111/144] Processing POWERGRID...


2026-01-11 12:46:46,832 - INFO - ✓ Fetched 662 records for POWERGRID


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/POWERGRID.csv

[112/144] Processing PVRINOX...


2026-01-11 12:46:47,701 - INFO - ✓ Fetched 662 records for PVRINOX


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/PVRINOX.csv

[113/144] Processing RAMCOCEM...


2026-01-11 12:46:48,547 - INFO - ✓ Fetched 662 records for RAMCOCEM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/RAMCOCEM.csv

[114/144] Processing RBLBANK...


2026-01-11 12:46:49,401 - INFO - ✓ Fetched 662 records for RBLBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/RBLBANK.csv

[115/144] Processing RECLTD...


2026-01-11 12:46:50,257 - INFO - ✓ Fetched 662 records for RECLTD


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/RECLTD.csv

[116/144] Processing RELIANCE...


2026-01-11 12:46:51,118 - INFO - ✓ Fetched 662 records for RELIANCE


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/RELIANCE.csv

[117/144] Processing SAIL...


2026-01-11 12:46:51,969 - INFO - ✓ Fetched 662 records for SAIL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SAIL.csv

[118/144] Processing SBIN...


2026-01-11 12:46:52,851 - INFO - ✓ Fetched 662 records for SBIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SBIN.csv

[119/144] Processing SHREECEM...


2026-01-11 12:46:53,717 - INFO - ✓ Fetched 662 records for SHREECEM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SHREECEM.csv

[120/144] Processing SIEMENS...


2026-01-11 12:46:54,573 - INFO - ✓ Fetched 662 records for SIEMENS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SIEMENS.csv

[121/144] Processing SRF...


2026-01-11 12:46:55,422 - INFO - ✓ Fetched 662 records for SRF


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SRF.csv

[122/144] Processing SHRIRAMFIN...


2026-01-11 12:46:56,295 - INFO - ✓ Fetched 662 records for SHRIRAMFIN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SHRIRAMFIN.csv

[123/144] Processing SUNPHARMA...


2026-01-11 12:46:57,144 - INFO - ✓ Fetched 662 records for SUNPHARMA


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SUNPHARMA.csv

[124/144] Processing SUNTV...


2026-01-11 12:46:57,999 - INFO - ✓ Fetched 662 records for SUNTV


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/SUNTV.csv

[125/144] Processing TATACHEM...


2026-01-11 12:46:58,861 - INFO - ✓ Fetched 662 records for TATACHEM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TATACHEM.csv

[126/144] Processing TATACONSUM...


2026-01-11 12:46:59,729 - INFO - ✓ Fetched 662 records for TATACONSUM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TATACONSUM.csv

[127/144] Processing TMPV...


2026-01-11 12:47:00,583 - INFO - ✓ Fetched 662 records for TMPV


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TMPV.csv

[128/144] Processing TATAPOWER...


2026-01-11 12:47:01,453 - INFO - ✓ Fetched 662 records for TATAPOWER


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TATAPOWER.csv

[129/144] Processing TATASTEEL...


2026-01-11 12:47:02,321 - INFO - ✓ Fetched 662 records for TATASTEEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TATASTEEL.csv

[130/144] Processing TCS...


2026-01-11 12:47:03,184 - INFO - ✓ Fetched 662 records for TCS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TCS.csv

[131/144] Processing TECHM...


2026-01-11 12:47:04,055 - INFO - ✓ Fetched 662 records for TECHM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TECHM.csv

[132/144] Processing TITAN...


2026-01-11 12:47:04,921 - INFO - ✓ Fetched 662 records for TITAN


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TITAN.csv

[133/144] Processing TORNTPHARM...


2026-01-11 12:47:05,771 - INFO - ✓ Fetched 662 records for TORNTPHARM


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TORNTPHARM.csv

[134/144] Processing TORNTPOWER...


2026-01-11 12:47:06,625 - INFO - ✓ Fetched 662 records for TORNTPOWER


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TORNTPOWER.csv

[135/144] Processing TVSMOTOR...


2026-01-11 12:47:07,485 - INFO - ✓ Fetched 662 records for TVSMOTOR


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/TVSMOTOR.csv

[136/144] Processing UBL...


2026-01-11 12:47:08,356 - INFO - ✓ Fetched 662 records for UBL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/UBL.csv

[137/144] Processing UJJIVANSFB...


2026-01-11 12:47:09,226 - INFO - ✓ Fetched 662 records for UJJIVANSFB


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/UJJIVANSFB.csv

[138/144] Processing ULTRACEMCO...


2026-01-11 12:47:10,083 - INFO - ✓ Fetched 662 records for ULTRACEMCO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ULTRACEMCO.csv

[139/144] Processing UPL...


2026-01-11 12:47:10,941 - INFO - ✓ Fetched 662 records for UPL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/UPL.csv

[140/144] Processing VEDL...


2026-01-11 12:47:11,806 - INFO - ✓ Fetched 662 records for VEDL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/VEDL.csv

[141/144] Processing VOLTAS...


2026-01-11 12:47:12,654 - INFO - ✓ Fetched 662 records for VOLTAS


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/VOLTAS.csv

[142/144] Processing WIPRO...


2026-01-11 12:47:13,521 - INFO - ✓ Fetched 662 records for WIPRO


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/WIPRO.csv

[143/144] Processing YESBANK...


2026-01-11 12:47:14,373 - INFO - ✓ Fetched 662 records for YESBANK


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/YESBANK.csv

[144/144] Processing ZEEL...


2026-01-11 12:47:15,228 - INFO - ✓ Fetched 662 records for ZEEL


✓ Saved 662 records to ./data/NIFTY200_Subset10/raw/ZEEL.csv

DOWNLOAD COMPLETE
Successful: 143/144
Failed: 1

Failed symbols: EQUITASBNK


## 12. Generate Quality Report

In [30]:
# Save quality report
report_file = os.path.join(OUTPUT_DIR, "data_quality_report.txt")
with open(report_file, 'w') as f:
    f.write("Phase 8: Data Quality Report\n")
    f.write("="*60 + "\n\n")
    f.write(f"Date Range: {FROM_DATE} to {TO_DATE}\n")
    f.write(f"Universe: F&O stocks as of Apr 2020\n")
    f.write(f"Total stocks in universe: {len(symbols)}\n")
    f.write(f"Successfully downloaded: {successful}\n")
    f.write(f"Failed/Skipped: {len(failed)}\n\n")
    
    f.write("Validation Summary:\n")
    f.write("-"*60 + "\n")
    
    # Count by quality status
    pass_count = sum(1 for v in validation_results if v['data_quality'] == 'PASS')
    warn_count = sum(1 for v in validation_results if v['data_quality'] == 'WARN')
    fail_count = sum(1 for v in validation_results if v['data_quality'] == 'FAIL')
    
    f.write(f"PASS: {pass_count}\n")
    f.write(f"WARN: {warn_count}\n")
    f.write(f"FAIL: {fail_count}\n\n")
    
    if failed:
        f.write(f"Failed symbols:\n{', '.join(failed)}\n")

print(f"\n✓ Quality report saved to: {report_file}")
print(f"\nSummary:")
print(f"  PASS: {pass_count}")
print(f"  WARN: {warn_count}")
print(f"  FAIL: {fail_count}")


✓ Quality report saved to: ./data/NIFTY200_Subset10/data_quality_report.txt

Summary:
  PASS: 142
  WARN: 1
  FAIL: 1
